Add SealingState; don't prepare block when not ready.

afck opened the pull request:
This prevents the miner from creating new blocks for sealing even if it's not our turn to seal them.

However, it likely does not detect all cases yet where it's not our turn.

(Closes #70.)

afck commented:
In my local test, it worked as expected, and it seemed to prevent _most_ of the cases where we prepared unsealable blocks, but not all.

This looks like a safe change to me, but I wonder why `seals_internally` wasn't doing this in the first place. Let me know if you see any potential problems with this!

DemiMarie commented:
Can we propose this upstream?

DemiMarie commented:
I suspect the problem is missing synchronization.

Can we place a global lock around the engine and miner?

afck commented:
Yes, I suspect that, too: Ideally, we _do_ want to start preparing a block, even if our turn has not quite started yet, so that when it starts, we're ready to seal it. So basically we want to prepare one in our predecessor's turn as soon as we've received our predecessor's block, i.e. it shouldn't depend on time at all, but only on the latest imported block. So maybe this PR is the wrong approach after all?

> Can we place a global lock around the engine and miner?

Somehow it seems in Parity every struct has all of its fields locked internally. It may be worth a try, but I suspect that global locks around engine and miner would cause them to deadlock, since there are places where (directly or indirectly) they call each other in both directions. :grimacing:

DemiMarie commented:
I think so.  I suspect that the real answer is that anything we need to do in a block we are sealing, needs to be done in the actual function that seals the block.

afck commented:
> anything we need to do in a block we are sealing, needs to be done in the actual function that seals the block.

So you're essentially saying the issue doesn't have a solution, and we always need to prepare the block and try to seal it, like we currently do?

afck commented:
I force-pushed a new approach that doesn't use `self.step` at all. (Not sure if it should.) Instead it only checks whether the block after the latest one is our turn.

vkomenda commented on ethcore/src/engines/authority_round/mod.rs:
```diff
@@ -226,9 +226,15 @@ impl EpochManager {
+		client: &EngineClient,
+		machine: &EthereumMachine,
+		validators: &ValidatorSet,
+		hash: H256
```
Does this slice need to be passed by value, or could you pass it by reference?

afck replied:
Unfortunately the `client.epoch_transition_for(hash)` call below takes it by value, so it would have to be copied there instead of moved.
(This way, if the caller of `zoom_to_after` also just moved the value, there's no copy at all.)

afck commented:
I also addressed the TODO: In `on_prepare_block`, we are always the block's author, so we don't have to take our address from the engine signer.

DemiMarie commented:
@afck In retrospect, what I am saying is: _`on_prepare_block` must be idempotent._

DemiMarie commented on ethcore/src/engines/authority_round/mod.rs:
```diff
@@ -1007,7 +1055,7 @@ impl Engine<EthereumMachine> for AuthorityRound {
 		}
 
 		let rlp = Rlp::new(rlp);
-		let empty_step: EmptyStep = rlp.as_val().map_err(fmt_err)?;;
```
How did this even compile?

DemiMarie commented on ethcore/src/engines/authority_round/mod.rs:
```diff
@@ -997,8 +1003,50 @@ impl Engine<EthereumMachine> for AuthorityRound {
+				debug!(target: "engine", "Not preparing block: Unable to zoom to epoch.");
+				return Some(false);
+			}
+			CowLike::Owned(epoch_manager.validators().clone())
```
This looks great.

DemiMarie commented on ethcore/src/engines/authority_round/mod.rs:
```diff
@@ -997,8 +1003,50 @@ impl Engine<EthereumMachine> for AuthorityRound {
+		};
+
+		let parent_step = header_step(&parent, self.empty_steps_transition)
+			.expect("Header has been verified; qed");
```
Could this fail due to an I/O error involving the database?

phahulin commented:
If I'm not mistaken, @DemiMarie 's remark can be extended to the case when at least one of the validators is down and skips blocks - then the network gets stuck because `step` is not advanced
```
2019-03-04 19:41:45  IO Worker #2 TRACE engine  Fetched proposer for step 310343469: 0xbbca…4c78
2019-03-04 19:41:45  IO Worker #2 TRACE engine  Not preparing block: not our turn. (Our address: 0x75df…9441)
...
2019-03-04 19:41:50  IO Worker #1 TRACE engine  Fetched proposer for step 310343469: 0xbbca…4c78
2019-03-04 19:41:50  IO Worker #1 TRACE engine  Not preparing block: not our turn. (Our address: 0x75df…9441)
...
2019-03-04 19:41:55  IO Worker #1 TRACE engine  Fetched proposer for step 310343469: 0xbbca…4c78
2019-03-04 19:41:55  IO Worker #1 TRACE engine  Not preparing block: not our turn. (Our address: 0x75df…9441)
```

afck commented:
> then the network gets stuck because `step` is not advanced

You're right! :weary: I don't have an idea yet how to solve this. But somehow `generate_seal` gets around that issue, so `seals_internally` should be able to do that, too…
I'll give it another try tomorrow.

DemiMarie commented:
@afck is it in `clear_empty_steps()`?

DemiMarie commented:
Fixed.  We were using the wrong `step` value.

DemiMarie commented:
@phahulin can you test this?

phahulin commented:
@DemiMarie couldn't get past the first few (1-2) blocks with all nodes up:
```
2019-03-05 11:00:20  IO Worker #3 TRACE miner  update_sealing
2019-03-05 11:00:20  IO Worker #3 TRACE miner  requires_reseal: sealing enabled
2019-03-05 11:00:20  IO Worker #3 TRACE engine  Fetched proposer for step 310354564: 0x75df…9441
2019-03-05 11:00:20  IO Worker #3 TRACE miner  requires_reseal: should_disable_sealing=false; forced=true, has_local=false, internal=Some(true), had_requests=false
2019-03-05 11:00:20  IO Worker #3 TRACE miner  update_sealing: preparing a block
2019-03-05 11:00:20  IO Worker #3 TRACE miner  prepare_block: Already have previous work; updating and returning
2019-03-05 11:00:20  IO Worker #3 DEBUG miner  Attempting to push 1 transactions.
2019-03-05 11:00:20  IO Worker #3 DEBUG miner  Adding tx 0xb214dc5e9ee5c97ff243c2832db03a0c5ca414ec545f09406f036b770447e181 took 50 ms
2019-03-05 11:00:20  IO Worker #3 DEBUG miner  Pushed 0 transactions in 51 ms
2019-03-05 11:00:20  IO Worker #3 TRACE reward  Block reward benefactors: [(0x75df42383afe6bf5194aa8fa0e9b3d5f9e869441, Author)]
2019-03-05 11:00:20  IO Worker #3 TRACE engine  Fetched proposer for step 310354564: 0x75df…9441
2019-03-05 11:00:20  IO Worker #3 TRACE miner  update_sealing: engine indicates internal sealing
2019-03-05 11:00:20  IO Worker #3 TRACE miner  seal_block_internally: attempting internal seal.
2019-03-05 11:00:20  IO Worker #3 DEBUG engine  Aborting seal generation. The step or empty_steps have changed in the meantime. 340282366920938463463374607431768211454 != 340282366920938463463374607431768211443
```
I've seen `Aborting seal... step or empty_steps have changed...` before, but now it happens every step.

afck commented:
I can confirm that it fails for me, too, both with and without the last commit, if I disable nodes 2 and 5.
With `aura-pos`, it works.

afck commented:
…but disabling the sealing queue (i.e. explicitly never taking [that branch](https://github.com/poanetwork/parity-ethereum/blob/b5097c67155debd9aee2c5e499456b4a1f187b26/ethcore/src/miner/miner.rs#L385)) fixes it for me!
I'll try to find out why we ended up there in the first place; I don't understand when that block was prepared and queued…

afck commented:
With [this change](https://github.com/poanetwork/parity-ethereum/pull/103/commits/c64e683c40e41c708a6905187931d3783f2d0fd4) it works for me.

Maybe `Engine` needs yet another getter so that we're able to either prevent [this `prepare_block` call](https://github.com/poanetwork/parity-ethereum/blob/c64e683c40e41c708a6905187931d3783f2d0fd4/ethcore/src/miner/miner.rs#L1107) or to return `None` from `prepare_block` itself if the engine isn't ready for sealing.

DemiMarie commented:
@afck It might, but it is probably a good idea to merge it as-is, unless @vkomenda or @phahulin say otherwise.

phahulin commented:
IMO we shouldn't merge it yet, because the issue itself is not a blocker, and it's not yet clear (at least to me), what can we break by disabling the sealing queue.

afck commented:
I agree; let's not merge something where we don't know what we're doing:
At least we should understand why this doesn't work with the sealing queue. Why doesn't the same problem already occur on `aura-pos` or `stable`…?

afck commented:
OK, I think `aura-pos` and `stable` also don't use the sealing queue!
* [This `set_pending` call](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L658) is never made because we never return a `Seal::Proposal`.
* [This one](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L720) never happens because `prepare_work` [is only called](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L1136) if `seals_internally` returns `None`, [or  in `prepare_pending_block`](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L785), which is [not called here](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L814) if `seals_internally` is always `Some(true)`, and [not called here](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L1152), because `work_package` is probably also never called with Aura…?
* And [this one](https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L1132) is never reached because `seals_internally` never returns `Some(false)`.

But I'm not sure yet what the proper solution is.

vkomenda commented:
@afck, do you think https://github.com/poanetwork/parity-ethereum/blob/85b58a91508dc42585389db58c8aea4347aff61f/ethcore/src/miner/miner.rs#L384 always evaluates to `None`? That should happen if the queue is not used.

afck commented:
Yes, I'm pretty sure it does, on `aura-pos` and with my WIP change.

afck commented:
I added a fourth value to `seals_internally` and renamed it to `sealing_state`. I _think_ this is doing the right thing now, but it needs to be carefully double-checked. :grimacing:

DemiMarie commented:
@afck `SealingState::Prepare` is dead code.  Do you plan to change that?

afck commented:
Good point… so maybe we don't need `SealingState` at all, and just need to change the effect of `Some(false)` to what's now `SealingState::None`… I'll have a closer look tomorrow.

DemiMarie commented on ethcore/src/miner/miner.rs:
```diff
@@ -1119,19 +1126,20 @@ impl miner::MinerService for Miner {
 				self.sealing.lock().queue.set_pending(block);
 			},
-			None => {
+			SealingState::None => (), // unreachable
```
If this really is unreachable, we should `panic!()`, or at least log something.

DemiMarie commented on ethcore/src/miner/miner.rs:
```diff
@@ -1119,19 +1126,20 @@ impl miner::MinerService for Miner {
 			},
-			Some(false) => {
-				trace!(target: "miner", "update_sealing: engine is not keen to seal internally right now");
+			SealingState::Prepare => {
```
This branch is currently dead code, as `SealingState::Prepare` is never constructed.

DemiMarie commented:
I think that we should keep the `SealingState` enum, since it is much more descriptive than `Option<bool>`.  Perhaps

```rust
enum SealingState {
    Ready,
    NotReady,
    External,
}
```
would be a good fit?

afck commented:
Thanks, I renamed the variants as you suggested, and removed `Prepare`.

I had originally meant `Prepare` to be what was `Some(false)` before, but it looks like `Some(false)` was only returned by some engines if there was no mining address. Since that means they can't ever seal a block anyway, I changed their return value to `NotReady`.

afck commented on ethcore/src/miner/miner.rs:
```diff
@@ -1127,18 +1127,13 @@ impl miner::MinerService for Miner {
-				self.sealing.lock().queue.set_pending(block);
-			},
-			SealingState::None => (), // unreachable
+			SealingState::NotReady => unreachable!("If None, we returned right after sealing_state was computed. qed."),
```
Right, thanks! That was redundant anyway. I updated the message.

afck commented:
I squashed and rebased.

varasev commented:
@afck Please clarify how this can be tested? If I add a node which is not a validator to `posdao-test-setup` and launch this, there shouldn't be any warnings in its logs, right? So, if I don't see warnings for such a node, can this PR be considered as successful?

In other words, how can I understand that these changes work fine?

afck commented:
I think it rather affects nodes that _are_ validators: Without this change, they print messages that they are not ready to seal, because they prepare blocks all the time, even if it's not their turn:
```
TRACE engine  generate_seal: 0xbbca…4c78 not a proposer for step 309905993
```
But for some reason they're not even warnings, just traces, even though it seems like a big waste of CPU time.
So that's the part that's easy to test: That those messages are gone.

The hard part is: Could this PR cause any other problems in practice? That's why I'd really like someone to review it who is _very_ familiar with the inner workings of Parity and Aura, and it's why I also made an upstream PR: https://github.com/paritytech/parity-ethereum/pull/10529
(But nobody has looked at it yet.)

varasev commented:
Ok, I left ping in that PR, hope the guys will help to check the code.

varasev commented:
Since Parity approved the same changes for upstream, I'll try to launch the build based on the `afck-seals-internally` branch and if it's ok, we will merge it.

varasev commented:
I tried to launch it and don't see the trace message `TRACE engine  generate_seal: 0xbbca…4c78 not a proposer for step ...` anymore. Also, the CPU time seems to be reduced 👍 So, I think we can merge it.
